tools/kagami/src/swarm.rs

```diff
@@ -0,0 +1,1314 @@
+    }
+}
+
+mod key_gen {
+    use iroha_crypto::{error::Error, KeyGenConfiguration, KeyPair};
+
+    /// If there is no base seed, the additional one will be ignored
+    pub fn generate(base_seed: Option<&[u8]>, additional_seed: &[u8]) -> Result<KeyPair, Error> {
+        let cfg = base_seed
+            .map(|base| {
+                let seed: Vec<_> = base.iter().chain(additional_seed).copied().collect();
+                KeyGenConfiguration::default().use_seed(seed)
+            })
+            .unwrap_or_default();
+
+        KeyPair::generate_with_configuration(cfg)
+    }
+}
+
+mod peer_generator {
+    use std::{collections::BTreeMap, num::NonZeroU16};
+
+    use color_eyre::{eyre::Context, Report};
+    use iroha_crypto::KeyPair;
+    use iroha_data_model::prelude::PeerId;
+    use iroha_primitives::addr::{SocketAddr, SocketAddrHost};
+
+    const BASE_PORT_P2P: u16 = 1337;
+    const BASE_PORT_API: u16 = 8080;
+    const BASE_PORT_TELEMETRY: u16 = 8180;
+    const BASE_SERVICE_NAME: &'_ str = "iroha";
+
+    pub struct Peer {
+        pub name: String,
+        pub port_p2p: u16,
+        pub port_api: u16,
+        pub port_telemetry: u16,
+        pub key_pair: KeyPair,
+    }
+
+    impl Peer {
+        pub fn id(&self) -> PeerId {
+            PeerId::new(&self.addr(self.port_p2p), self.key_pair.public_key())
+        }
+
+        pub fn addr(&self, port: u16) -> SocketAddr {
+            SocketAddr::Host(SocketAddrHost {
+                host: self.name.clone().into(),
+                port,
+            })
+        }
+    }
+
+    pub fn generate_peers(
+        peers: NonZeroU16,
+        base_seed: Option<&[u8]>,
+    ) -> Result<BTreeMap<String, Peer>, Report> {
+        (0u16..peers.get())
+            .map(|i| {
+                let service_name = format!("{BASE_SERVICE_NAME}{i}");
+
+                let key_pair = super::key_gen::generate(base_seed, service_name.as_bytes())
+                    .wrap_err("failed to generate key pair")?;
+
+                let peer = Peer {
+                    name: service_name.clone(),
+                    port_p2p: BASE_PORT_P2P + i,
+                    port_api: BASE_PORT_API + i,
+                    port_telemetry: BASE_PORT_TELEMETRY + i,
+                    key_pair,
+                };
+
+                Ok((service_name, peer))
+            })
+            .collect()
+    }
+}
+
+mod serialize_docker_compose {
+    use std::{
+        collections::{BTreeMap, BTreeSet},
+        fmt::Display,
+        fs::File,
+        io::Write,
+        path::PathBuf,
+    };
+
+    use color_eyre::eyre::{eyre, Context};
+    use iroha_crypto::{KeyPair, PrivateKey, PublicKey};
+    use iroha_data_model::prelude::PeerId;
+    use iroha_primitives::addr::SocketAddr;
+    use serde::{ser::Error as _, Serialize, Serializer};
+
+    use crate::swarm::peer_generator::Peer;
+
+    const COMMAND_SUBMIT_GENESIS: &str = "iroha --submit-genesis";
+    const DOCKER_COMPOSE_VERSION: &str = "3.8";
```

Comment:
Can we fetch it somehow instead of hard-coding? I feel it could be the first thing to go out of sync.

Reply:
The first thing to go out of sync is configuration environment variables, for sure. Hopefully, unit tests will prevent it.

As for `iroha --submit-genesis` - it could become obsolete if the main CLI changes, or there are changes in Dockerfile. I hope that a person who will update them will also bother about updating the mechanism that generates Docker Compose files. 

Anyway, it would be nice to fetch the command from `Dockerfile` itself, or from the CLI code, but... Idk, it seems to overcomplication - just change a single line which changes rarely.

As for Docker Compose version, how could it become obsolete without refactoring the swarm itself?

Probably I miss some understanding. Is there any source it could be fetched from, and is there reason to do so?

Reply:
NVM, was mostly thinking along the lines of overcomplication. Have no complaints otherwise.